SMAPIv3 `zfs-vol` driver (aka zfs-ng gen2)

ydirson opened the pull request:
This is a 2nd gen prototype based on #11, incrementally rebuilt piece by piece

gthvn1 commented on plugins/volume/org.xen.xapi.storage.zfs-ng/sr.py:
```diff
@@ -0,0 +1,133 @@
+        # - from a "devices" config string (comma-separated list of devices)
+
+        if 'zpool' in configuration:
+            if 'devices' not in configuration:
```
it is if 'devices' in configuration I think

ydirson replied:
yes, that was corrected in a later fixup - everything has been squashed now

gthvn1 commented:
This first part looks good :+1:

ydirson committed:
zfs-vol: implement sr-create 1/4: SR.create

Pool passed through `zpool` should still get a sanity check before
being accepted, and get aligned with what we otherwise create, to be
completed later.

SR can then be disposed of using "xe sr-forget" and "zpool destroy".

Underlying zpool can be either separately created and given to
SR.create, or a set of devices can be specified to create a new zpool
on.

Notes:
- `SR.attach` will be necessary to complete `sr-create`: while from a
  calling user PoV things appear to have gone fine, SMlog reveals a
  call to `SR.attach`, which naturally failed as not yet implemented.
  Any attempt to `xe pbd-plug` naturally also fails.  See
  xapi-project/xen-api#5532.
- there is no plugin feature to shield from SR.* calls

Originally-by: Matias Ezequiel Vara Larsen <matias.vara@vates.fr>
Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

ydirson committed:
zfs-vol: base infrastructure

Bare minimum for a plugin, no operation implemented.

Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

gthvn1 commented on plugins/volume/org.xen.xapi.storage.zfs-ng/zfsutils.py:
```diff
@@ -0,0 +1,38 @@
+    return call(dbg, cmd).strip()
+
+def pool_create(dbg, pool_name, devs):
+    cmd = [
```
nit and it is ok for me but why not splitting cmd like others?

ydirson replied:
I pondered that, but I felt it as more clear to keep the options together with their arguments, which only left "zpool create" to split, so that would look quite weird.  But then I left the `.split()` for `destroy` so yes it is in fact lacking consistency

gthvn1 commented on plugins/volume/org.xen.xapi.storage.zfs-ng/zfsutils.py:
```diff
@@ -0,0 +1,38 @@
+    return int(call(dbg, cmd))
+
+def vdi_create(dbg, zvol_path, size_mib):
+    cmd = [ 'zfs', 'create',
```
same here

ydirson committed:
zfs-vol: implement Volume.destroy

Originally-by: Matias Ezequiel Vara Larsen <matias.vara@vates.fr>
Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

ydirson committed:
zfs-vol: implement Volume.resize

mimics libcow to:
- forbid shrinking
- temporarily write None in the DB to avoid discrepancy between
  (cached) vsize in DB and reality when a crash occurs while resizing

do better than libqcow by:
- updating vdi.volume.vsize to match the db *on resize too*, and *on
  create too*, logging the actual size applied
- calling _vdi_sanitize everywhere we load the vdi from db

Originally-by: Guillaume Thouvenin <guillaume.thouvenin@vates.tech>
Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

ydirson committed:
zfs-vol: activate default implementation for Volume.set_name and set_description

Originally-by: Matias Ezequiel Vara Larsen <matias.vara@vates.fr>
Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

ydirson committed:
zfs-vol: implement Volume.snapshot 1/n

Snapshot creation

Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

ydirson committed:
zfs-vol: implement Volume.snapshot 2/n

SR.ls for `xe sr-scan`

Originally-by: Matias Ezequiel Vara Larsen <matias.vara@vates.fr>
Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

ydirson committed:
zfs-vol: implement Volume.snapshot 3/n

Volume.stat e.g. for `xe vdi.destroy` or `xe vdi-copy into-vdi-uuid=`

`xe vdi-copy` then assumes we have a blockdev available for the
snapshot, which is wrong - we'd have to either stream the data or
create a clone volume.

Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

ydirson committed:
zfs-vol: implement Volume.clone 1/n

"Clone a snapshot" only, especially needed for `xe snapshot-revert`.

Originally-by: Matias Ezequiel Vara Larsen <matias.vara@vates.fr>
Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

gthvn1 commented on plugins/volume/org.xen.xapi.storage.zfs-ng/zfsutils.py:
```diff
@@ -22,12 +22,23 @@ def zvol_find_snap_path(dbg, pool_name, snap_id):
+    # a zvol depends on a snapshot if its origin is == to that
+    # snapshot, and depends on another zvol if its origin is a
+    # snapshot of that zvol
+    if "@" in vol_name:
```
"@" in the name means that it is a snapshot right? So you are speaking about snapshot that depends of another snapshots here?

ydirson replied:
no, this code checks whether other zvols depend on `vol_name`, and the check is different whether  `vol_name` is a snapshot (will appear directly as origin) or a zvol (where we look for snapshots based on it)

ydirson replied:
code disappeared since

ydirson committed:
zfs-vol: implement Volume.clone 2/n

Allow destroying a VDI when it has been cloned, by first "zfs-promoting"
each of those clones to remove all dependencies on it.

Makes sure to log the pool state before and after promotions, to be able
to understand any problem with this logic.

Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

ydirson committed:
zfs-vol: implement Volume.snapshot 4/n

Allow destruction of a snapshot with no dependency.

Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

ydirson committed:
zfs-vol: fork xapi.storage.common.call() to shoehorse retry-on-busy

This is obviously not a long-term solution, we should transition to call
into the official zfs python lib instead of spawning processes.

Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

ydirson committed:
zfs-vol: activate retry on busy on commands doing write operations

Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

ydirson committed:
zfs-vol: implement Volume.snapshot 5/n

When a snapshot with no child clones would block volume destruction,
create such a child clone.

Nothing will ever attempt to destroy such child clones, some sort of
GC is required.

Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

ydirson committed:
metabase: log VDI creations

This gives us a single log where all VDIs are registered, to get the
UUID / volume_id mapping for quicker debugging.

Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

ydirson committed:
zfs-vol: implement sr-create 3/4: store pool name in meta.json

This will be required by `SR.stat`, which is needed by `xe sr-plug`.

Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

ydirson committed:
zfs-vol: implement sr-create 4/4: SR.stat

This allows the "PDB plug on create" to work and completes the `xe
sr-create` implementation.

Note that since we cannot unplug a PBD yet the "xe sr-forget && zpool
destroy" we used up till now cannot work any more until that part is
done.

This is also another mandatory method needed by "xe sr-destroy".

Originally-by: Matias Ezequiel Vara Larsen <matias.vara@vates.fr>
Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

gthvn1 committed:
zfs-vol: implement sr-create 2/4: SR.attach

Import and mount the zpool if not mounted.

This is the first mandatory method needed by `xe sr-plug` and
`xe sr-destroy`.

`xe sr-plug` still does not complete, as it makes an `SR.stat` call
afterwards.

Originally-by: Matias Ezequiel Vara Larsen <matias.vara@vates.fr>
Originally-by: Yann Dirson <yann.dirson@vates.tech>
Signed-off-by: Guillaume <guillaume.thouvenin@vates.tech>

gthvn1 committed:
zfs-vol: implement sr-unplug: SR.detach

Export the zpool when detaching the SR.

This is also another mandatory method needed by `xe sr-destroy`.

Originally-by: Matias Ezequiel Vara Larsen <matias.vara@vates.fr>
Originally-by: Yann Dirson <yann.dirson@vates.tech>
Signed-off-by: Guillaume <guillaume.thouvenin@vates.tech>

ydirson committed:
zfs-vol: implement sr-destroy: SR.destroy

Essentials of the SR lifecycle is now done.

Originally-by: Matias Ezequiel Vara Larsen <matias.vara@vates.fr>
Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

ydirson committed:
zfs-vol: implement sr-param-set: add SR.set_name and SR.set_description

Originally-by: Matias Ezequiel Vara Larsen <matias.vara@vates.fr>
Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

ydirson committed:
zfs-vol: implement volume-create 1/2: Volume.set and .unset

This activates the default implementations from libcow (insertion in
vdi_custom_keys table in SR metabase).  `Volume.set` is necessary for
`xe vdi-create`, and `Volume.unset` is provided for symmetry.

Originally-by: Matias Ezequiel Vara Larsen <matias.vara@vates.fr>
Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

ydirson committed:
zfs-vol: implement volume-create 2/2: Volume.create

We create sparse volumes, making the SR a thin-provisioning one.

Cache as vsize the actual size created by the zfs command (which in
zfs-2.1 unlike in zfs-0.8 does round up requested size to be acceptable).

Note that if Volume.set were not implemented yet (or indeed cannot
complete and returns an error), we face xapi-project/xen-api#5530: if
the Volume.set call fails (to add the vdi-type=user custom-key
record), then the VDI object is not created (though, as this happens
after Volume.create, the zvol and DB entries were indeed created and
there suppression was not requested by SMAPI after the error).  But
since all the VDI information that count are here, "xe sr-scan" will
bring those VDI to life (after implementation of SR.ls).

Originally-by: Matias Ezequiel Vara Larsen <matias.vara@vates.fr>
Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

ydirson committed:
zfs-vol: implement sr-scan: SR.ls

Originally-by: Matias Ezequiel Vara Larsen <matias.vara@vates.fr>
Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

ydirson committed:
zfs-vol: implement Volume.stat

This is necessary e.g. for `xe vdi-copy into-vdi-uuid=`

Originally-by: Matias Ezequiel Vara Larsen <matias.vara@vates.fr>
Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

ydirson committed:
zfs-vol: expose device path for access to data

This is needed volume copy, `Datapath.attach` for `xe vbd-plug`, and such.

Originally-by: Matias Ezequiel Vara Larsen <matias.vara@vates.fr>
Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

gthvn1 commented on plugins/volume/org.xen.xapi.storage.zfs-ng/plugin.py:
```diff
@@ -0,0 +1,42 @@
+            "version": "3.0",
+            "required_api_version": "5.0",
+            "features": [
+                "VDI_CREATE",
```
shouldn't we add *SR_* features as well?

ydirson replied:
I deliberately did not add those whose for which I could not see any impact omitting them.
I must say I have trouble reading the ocaml code checking for features, which from an orbital bird's view looks like it should check them.

gthvn1 commented on plugins/volume/org.xen.xapi.storage.zfs-ng/sr.py:
```diff
@@ -0,0 +1,193 @@
+                log.error('"zpool" specified, "devices" should not be used')
+                raise Exception('"zpool" specified, "devices" should not be used')
+
+            # FIXME do we reject pools not under MOUNT_ROOT?
```
With #18 we choose where we are mounting the pool so we don't need to reject pools not under MOUNT_ROOT because we will force the mount point. I think to pool created with legacy mode by a user. We can deal with that.

ydirson replied:
I'd leave the FIXME's to be removed by #18, so things will be clear later when reading history

gthvn1 commented on plugins/volume/org.xen.xapi.storage.zfs-ng/sr.py:
```diff
@@ -0,0 +1,193 @@
+            # FIXME do we reject pools not under MOUNT_ROOT?
+
+            # FIXME validate existence of pool first?
+            pool_name = configuration['zpool']
```
Maybe it can be a good idea. But if we don't we will have an issue later when trying to attach the SR. So is it better to check now that we can import and mount the pool instead of do it later in the SR.attach. Maybe yes. We will need #18 for that :)

ydirson replied:
That somewhat converges with my comment on #18 about having a helper to know a zpool is already imported: that one would need its own helper to determine if a zpool exists even if not imported - or we can just define this mode of operation to mean "an imported zpool".

There is also the question whether we protect ourselves against getting passed a zpool what would already be in use, which OTOH expecting "an exported zpool" would answer.

gthvn1 commented on plugins/volume/org.xen.xapi.storage.zfs-ng/sr.py:
```diff
@@ -0,0 +1,193 @@
+            log.error('devices config must have "zpool" or "devices"')
+            raise Exception('devices config must have "zpool" or "devices"')
+
+        # FIXME this assumes zpool is mounted/attached
```
Again with #18 we can try import/mount. And in fact it answers to the previous FIXME where we finally need to check that the zpool provided by the user exist and can be mounted.

gthvn1 commented on plugins/volume/org.xen.xapi.storage.zfs-ng/sr.py:
```diff
@@ -0,0 +1,193 @@
+    def attach(self, dbg, configuration):
+        log.debug('{}: SR.attach: config={}'.format(dbg, configuration))
+
+        # ZFS automagically attaches a pool to a mountpoint on
```
fixed with #18

gthvn1 commented on plugins/volume/org.xen.xapi.storage.zfs-ng/zfsutils.py:
```diff
@@ -0,0 +1,132 @@
+        if error and proc.returncode != expRc:
+            log.error('%s: %s exitted with code %s: %s',
+                      dbg, " ".join(cmd_args), proc.returncode, stderr)
+            if ntries > 1 and "is busy" in stderr:
```
if returncode is 2 it means that the command line is invalid so we probably don't need to retry for this one.

ydirson replied:
fixed

gthvn1 commented on plugins/volume/org.xen.xapi.storage.zfs-ng/zfsutils.py:
```diff
@@ -0,0 +1,132 @@
+    return int(call(dbg, cmd))
+
+def vol_create(dbg, zvol_path, size_mib):
+    cmd = ("zfs create".split() + [zvol_path]
```
for my info why parenthesis are needed here?

ydirson replied:
python needs to know the expression continues on next line

gthvn1 reviewed (commented):
Still need to review that part about destroying snapshots. WIP...

gthvn1 commented on plugins/volume/org.xen.xapi.storage.zfs-ng/volume.py:
```diff
@@ -0,0 +1,270 @@
+        pool_name = meta["zpool"]
+
+        with VolumeContext(self.callbacks, sr, 'w') as opq:
+            # FIXME how should we choose image format?
```
RAW seems appropriate here. Do you think about adding an new format like BDEV (Block Device) ?

ydirson replied:
Yes RAW seems good, finally, as we've ruled out the nested-cow use-case anyway.

gthvn1 commented on plugins/volume/org.xen.xapi.storage.zfs-ng/volume.py:
```diff
@@ -0,0 +1,270 @@
+                    is_snapshot = vdi.volume.snap
+                    if is_snapshot:
+                        snap_name = zfsutils.zvol_find_snap_path(dbg, pool_name, vdi.volume.id)
+                        zfsutils.zpool_log_state(dbg, "before destroy {}".format(snap_name), pool_name)
```
nit: I think it is "before destroying"

ydirson replied:
in my mind it is "before `destroy` operation" ;)

gthvn1 commented:
I tested the creation of a VDI, snaphost it, clone it and delete the inital VDI. It seems ok. while testing I followed the code and it looks good to me :+1:

ydirson committed:
zfs-vol: new mode of SR creation with a ZFS vdev specification

This will allow management layers to present a suitable UI for the cases
they see fit without having to modify the interface.

Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

ydirson committed:
install: disable plugins not supposed to work on 8.3 currently

Signed-off-by: Yann Dirson <yann.dirson@vates.tech>

gthvn1 commented on plugins/volume/org.xen.xapi.storage.zfs-vol/sr.py:
```diff
@@ -0,0 +1,213 @@
+            log.error('devices config must have "zpool", "vdev", or "device"')
+            raise Exception('devices config must have "zpool", "vdev", or "device"')
+
+        # FIXME this assumes zpool is mounted/attached
```
maybe we can call attach (with version #18) ?

ydirson replied:
I'm not a big fan of the idea - we could maybe introduce a couple of helpers to be called by both, though.
That continues to advocate for #18 to become "squash" commits at some point

gthvn1 replied:
Sure for #18, you can take it to be part of your fixes (if this is what you mean). I will close it as implemented elsewhere. And I agree that calling attach is not great and probably adding an helper like "ensure_mounted" that tries to get the mountpoint and if not tries to import it once is better. Or also modified `import` to check if it is already mounted and let `import` be idempotent. In this case we can call it from here to ensure that it is mounted and raise an exception if it is not mounted and cannot be mounted.

gthvn1 commented on plugins/volume/org.xen.xapi.storage.zfs-vol/volume.py:
```diff
@@ -0,0 +1,301 @@
+            vdi_uri = self.callbacks.getVolumeUriPrefix(opq) + vdi_uuid
+
+        return {
+            'key': vdi_uuid,    # FIXME check this
```
the only requirement is that it should be unique? At least the common volume.py is doing the same. But I don't know if it is ok.

ydirson replied:
yes, does not look useful to dig deeper at this stage

gthvn1 commented on plugins/volume/org.xen.xapi.storage.zfs-vol/sr.py:
```diff
@@ -0,0 +1,213 @@
+
+            for vdi in vdis:
+                # TODO: handle this better
+                # _vdi_sanitize(vdi, opq, db, cb)
```
Shouldn't we call it instead of commenting? If we need to handle this better maybe it should be done in another patch that will modify `_vdi_sanitize`?

ydirson replied:
It's not that easy as this function is not applicable, but something similar is part of last weeks work still in fixup commits, see ab2dc908966e6724ab8179eeed5271113314d245 and later

ydirson replied:
My bad, that one being in `sr.py` not `volume.py` I missed it.  Addressed in upcoming push.

gthvn1 commented on plugins/volume/org.xen.xapi.storage.zfs-vol/volume.py:
```diff
@@ -0,0 +1,301 @@
+            'name': name,
+            'description': description,
+            'read_write': True,
+            'virtual_size': size,
```
shouldn't be `volume.vsize` that seems to be the adjusted size?

ydirson replied:
right, good catch!

ydirson replied:
fixed

ydirson commented:
Rerolled:
- applied fixups
- sync'd commit messages for new `zls-vol` name
- split #18 commit into `SR.attach` and `SR.detach` ones

gthvn1 commented:
LGTM

ydirson commented:
Pushed:
- removal of FIXME tags for `key` and `psize`
- fix of `psize` on `Volume.create` (we're creating thin volumes now)
- rewording of plugin description

gthvn1 commented on plugins/volume/org.xen.xapi.storage.zfs-vol/sr.py:
```diff
@@ -0,0 +1,218 @@
+                log.error('"zpool" specified, "device" or "vdev" should not be used')
+                raise Exception('"zpool" specified, "device" or "vdev" should not be used')
+
+            # FIXME validate existence of pool first?
```
I can propose a patch for this one based on zfs import maybe. So we can merge your patch and fix this later.

ydirson replied:
Post-proto work item

gthvn1 commented on plugins/volume/org.xen.xapi.storage.zfs-vol/sr.py:
```diff
@@ -0,0 +1,218 @@
+            raise Exception('devices config must have "zpool", "vdev", or "device"')
+
+        # FIXME this assumes zpool is mounted/attached
+        mountpoint = zfsutils.pool_mountpoint(dbg, pool_name)
```
maybe at least try expect and print an easy to understand error "expected zpool to be mounted" before raising the error? I'm not sure if raise directly is easy to read in log.

ydirson replied:
For the prototype it should be enough to get the limitation listed in the doc.  We have a work item for this post-proto.

gthvn1 reviewed (approved):
minor comments. LGTM !

ydirson commented:
New update with:
- avoid single-use temp psize/fsize variables
- remove remaining useless FIXME's, replaced by better info
- fixed some scoping errors with statements outside of the `with` blocks they belong too (python is too permissive) (but no runtime impact today)
- renamed and moved `zfsvol_vdi_sanitize` to `zfsutils` to add the missing call in `sr.ls`
- don't fail a `sr-scan` when we cannot locate zvol for a snapshot, just skip it in the output, and log an error

gthvn1 commented:
As the fixme about the zpool already mounted will be done later it seems reasonable for me to merge your PR.
Great job :+1:
